enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,891 @@
+This is technically very challenging and has been dismissed as an alternative.
+
+Documentation will be provided to explain to users the expected behaviours and the expectations to allow them to migrate their infrastructure across.
+
+## Design Details
+
+### Open Questions
+
+N/A
+
+## Test Plan
+
+### Conversion of provider specs
+
+The conversion of provider specs from Machine API to Cluster API should be able to be statically tested. We will build out extensive conversion testing via unit tests to ensure that, with a desired input, the correct conversion happens for the provider specs and the associated resources.
+
+### Authoritative API reconciliation
+
+Machine API and Cluster API controllers already use envtest as a way to provide integration testing with a "real" API server.
+We can extend the tests here to account for the cases where the `authoritativeAPI` feature is in use.
+
+By extending the integration tests, we can sculpt the desired inputs for mirror resources existing, not existing, and authoritative in both API versions,
+and then determine whether the reconcile continued or exited based on these conditions.
+
+### Full migration E2E
+
+The behaviours outlined [above][summary-of-rules] will be tested in a new E2E test suite specifically targeted at the conversion logic.
+
+In addition to these tests, the following tests will provide a general coverage of the conversion logic and behaviours defined:
+- Scale up and down of a mirrored MachineSet
+  - Copy an existing MachineSet in Machine API
+  - Observe that the new MachineSet is mirrored in Cluster API
+  - Switch the authoritative API to Cluster API
+  - Scale up the MachineSet in Cluster API
+  - Wait for Machine and mirror Machine to be created
+  - Wait for Node to join cluster
+  - Scale down MachineSet in Cluster API
+  - Observe scale down successfully
+  - Remove new MachineSet
+- Migration of ControlPlaneMachineSet to Cluster API
+  - Observe existing cluster ControlPlaneMachineSet
+  - Switch ControlPlaneMachineSet authoritative API to Cluster API
+  - Create a new InfrastructureTemplate with a larger instance size
+  - Update Cluster API ControlPlaneMachineSet to point to new InfrastructureTemplate
+  - Observe rolling update for ControlPlaneMachineSet
+- MachineHealthCheck in mixed clusters
+  - Ensure a mixed cluster is created with authoritative MachineSets in both APIs
+  - Trigger a MachineHealthCheck reconciliation using a Machine API authoritative MachineHealthCheck
+  - Ensure that Machines are deleted correctly
+  - Switch authority of MachineHealthCheck to Cluster API
+  - Trigger a second MachineHealthCheck reconciliation using the now Cluster API authoritative MachineHealthCheck
+  - Ensure that Machines are deleted correctly
+
+[summary-of-rules]: #summary-of-the-rules-outlined-above
+
+## Graduation Criteria
+
+The project will initially be introduced under a feature gate `MachineAPIMigration`.
```

Comment:
Do we need two feature gates? One for synchronization, one for enabling the authoritative API management controller?

Reply:
I had intended initially to use one gate to cover the whole feature, in the future though I expect we may have per platform gates as we evolve some of the migrations at a different pace.